src/main/java/com/google/googleidentity/oauth2/token/InMemoryOAuth2TokenService.java

```diff
@@ -0,0 +1,451 @@
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+package com.google.googleidentity.oauth2.token;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.io.BaseEncoding;
+import com.google.googleidentity.oauth2.request.OAuth2Request;
+import com.google.inject.Singleton;
+
+import javax.crypto.Cipher;
+import javax.crypto.KeyGenerator;
+import javax.crypto.NoSuchPaddingException;
+import java.security.InvalidParameterException;
+import java.security.Key;
+import java.security.NoSuchAlgorithmException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.HashSet;
+import java.util.UUID;
+import java.util.Date;
+import java.util.Optional;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Default in memory token service.
+ * An implementation of {@link OAuth2TokenService}.
+ * The encryption of clientId and username is only needed in this InMemory Design.
+ */
+@Singleton
+public class InMemoryOAuth2TokenService implements OAuth2TokenService {
+
+    private static final Logger log = Logger.getLogger("InMemoryOAuth2TokenService");
+
+    private long tokenValidTime = 10 * 60l;
+
+    Map<String, HashMap<String, HashMap<String, OAuth2AccessToken>>> userClientAccessTokenMap =
+            new ConcurrentHashMap<>();
+
+    Map<String, HashMap<String, OAuth2RefreshToken>> userClientRefreshTokenMap =
+            new ConcurrentHashMap<>();
+
+    private Key aesKey;
+
+    ScheduledExecutorService service;
+
+    public InMemoryOAuth2TokenService() {
+        initKey();
+        setTokenCleaner();
+    }
+
+    public InMemoryOAuth2TokenService(long tokenValidTime) {
+        this.tokenValidTime = tokenValidTime;
+        initKey();
+        setTokenCleaner();
+    }
+
+    /**
+     * Used for encrypt username and clientID
+     */
+    private class UserClientTokenInfo {
+        private String username;
+        private String clientID;
+        private String tokenValue;
+        private String tokenString;
+
+        private String delimiter = "\t";
+
+        UserClientTokenInfo(String username, String clientID, String tokenValue) {
+            this.username = username;
+            this.clientID = clientID;
+            this.tokenValue = tokenValue;
+            encryptTokenString();
+        }
+
+        UserClientTokenInfo(String tokenString) throws InvalidParameterException {
```

Comment:
Same here re: constructors doing work. I would recommend having a static method like UserClientTokenInfo parseTokenString(String tokenString) that ultimately invokes UserClientTokenInfo(username, clientID, tokenValue) once decryption and parsing is done.